### PR TITLE
feat: Script to scrape benchmark results into line protocol and automate gathering numbers

### PR DESCRIPTION
**Pull request by alamb**
Related to #574

This (relatively dirty) script scrapes the output of running `cargo bench --all` and turns it into line protocol. This is part of a larger plan of mine to gather and maintain historical performance numbers. 

Here is an example of it in action (the timestamps are the timestamps of the commit):
```
./benches/bench_scraper.sh /Users/alamb/Software/influxdb_iox

bench,gitsha=02fbce7ef7485bbbef67d6524d789a178f2ba559,hostname=ip-10-0-0-49.ec2.internal,group_name=integer_encode_random,bench_name=50 mean=347.31135171618814,mean_standard_error=2.7351040240461466,median=339.5733687494297,median_standard_error=2.635747931745169 1609190816000000000
bench,gitsha=02fbce7ef7485bbbef67d6524d789a178f2ba559,hostname=ip-10-0-0-49.ec2.internal,group_name=integer_encode_random,bench_name=10000 mean=69588.14785986746,mean_standard_error=359.2153907656142,median=68337.45194204859,median_standard_error=354.8700994103787 1609190816000000000
bench,gitsha=02fbce7ef7485bbbef67d6524d789a178f2ba559,hostname=ip-10-0-0-49.ec2.internal,group_name=integer_encode_random,bench_name=50000 mean=356778.424610302,mean_standard_error=2410.449744056784,median=351522.30402930407,median_standard_error=1903.5165168437857 1609190816000000000
bench,gitsha=02fbce7ef7485bbbef67d6524d789a178f2ba559,hostname=ip-10-0-0-49.ec2.internal,group_name=integer_encode_random,bench_name=250 mean=1375.179972827231,mean_standard_error=7.675859374037711,median=1363.2371625615765,median_standard_error=4.820258840010316 1609190816000000000

```

I plant to use this script to start gathering (and loading) historical performance runs into the tools cluster for display. I think the same basic idea can be used for arrow too.

Here is an example (created using `git rev-list main | head | sed -e 's|^|./run_bench.sh /Users/alamb/Software/influxdb_iox2 |'`) of the types of commands:

```
./run_bench.sh /Users/alamb/Software/influxdb_iox2 02fbce7ef7485bbbef67d6524d789a178f2ba559
./run_bench.sh /Users/alamb/Software/influxdb_iox2 937442cfa030b47396650e0eca2ced86417359bd
./run_bench.sh /Users/alamb/Software/influxdb_iox2 c46cf6fdcf1223c265958a929de049925fdffed9
./run_bench.sh /Users/alamb/Software/influxdb_iox2 c0dc93a8cb595e49590a0fc33cf527c5eca6ef71
./run_bench.sh /Users/alamb/Software/influxdb_iox2 0d0ec0ce69451f76ece559ac0271395172109621
./run_bench.sh /Users/alamb/Software/influxdb_iox2 5fa77c32cc6b685033b9cdc07058644218a8bf5b
./run_bench.sh /Users/alamb/Software/influxdb_iox2 c4952fc58d8baa8745065dc73f9ec4a64f199082
./run_bench.sh /Users/alamb/Software/influxdb_iox2 b998a916b3a7d566b0b02181bedbce5b03f5b444
./run_bench.sh /Users/alamb/Software/influxdb_iox2 1d200c5c77698275628762dc9d5eeafe4c2ea11b
./run_bench.sh /Users/alamb/Software/influxdb_iox2 f7febf3797adfc10211ffa04283ac90b77bc078b
```

And then the output iox_perf.lp can be loaded into the tools cluster (or influxdb_v2)

### CODE REVIEW

File: benches/scrape_benches.sh [hunk 1]
```diff
@@ -0,0 +1,109 @@
...
+    bench_name=${fields[1]/ /_}
+
+    hostname=`hostname`
+    echo -n "bench,gitsha=$GITSHA,hostname=${hostname},group_name=$group_name,bench_name=$bench_name "
+
+    # if this was done in a single jq command it would definitely be faster...
+    echo -n "mean="`jq '.mean.point_estimate' "$estimates_file"`
+    echo -n ",mean_standard_error="`jq '.mean.standard_error' "$estimates_file"`
+    echo -n ",median="`jq '.median.point_estimate' "$estimates_file"`
+    echo -n ",median_standard_error="`jq '.median.standard_error' "$estimates_file"`
```
Comment (carols10cents) [inline benches/scrape_benches.sh:103]: I *think* this works? Check me though:

```suggestion
    echo -n `jq  -j '"mean=" + (.mean.point_estimate | tostring), ",mean_standard_error=" + (.mean.standard_error | tostring), ",median=" + (.median.point_estimate | tostring), ",median_standard_error=" + (.median.standard_error | tostring)' "$estimates_file"`
```

**Review by carols10cents — DISMISSED** [review]
I'm farrr from a bash expert, but this looks generally fine? Some small suggestions.

File: benches/scrape_benches.sh [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+# Scrape data:
+#  ./bench_scraper /Users/alamb/Software/influxdb_iox2
+#
+# To load to
+
+
+function main {
+    SOURCE_DIR=$1
+    if [ -z "$SOURCE_DIR" ] ; then
+        echo "Error: target directory not specified"
```
[the remark this answers is not in the document]
Comment (alamb) [reply benches/scrape_benches.sh:17]: the source dir --good call

File: benches/scrape_benches.sh (hunk 1 above)
Comment (alamb) [reply benches/scrape_benches.sh:103]: It worked like a charm. I tested using

```
./benches/./scrape_benches.sh /Users/alamb/Software/influxdb_iox2
```

On my local system and the output was the same in both cases. 👍 

And the script is noticeably faster. Thanks @carols10cents

File: benches/run_bench.sh [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+Log is written to stdout
+Performance results are appended to iox_bench.lp
+EOF
+
+# Location of this script (assumes that
+# https://stackoverflow.com/questions/59895/how-to-get-the-source-directory-of-a-bash-script-from-within-the-script-itself
+SCRIPT_DIR="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
+
+SOURCE_DIR=$1
+if [ -z "$SOURCE_DIR" ] ; then
```
Comment (domodwyer) [inline benches/run_bench.sh:32]: Maybe `-d` here to bail if the dir doesn't exist?

**Review by domodwyer — APPROVED** [review]
Looks cool - beautifully commented!

File: benches/run_bench.sh (hunk 3 above)
Comment (alamb) [reply benches/run_bench.sh:32]: Good call -- added: 

```
alamb@MacBook-Pro influxdb_iox % ./benches/run_bench.sh /fooooo
./benches/run_bench.sh /fooooo
Error: Not a directory: /fooooo
Usage:
   ./benches/run_bench.sh <source directory> gitsha

Example:
   ./benches/run_bench.sh /Users/alamb/Software/influxdb_iox2 a6ed8d59dc0465ff3a605b1fd4783faa9edb8560

Log is written to stdout
Performance results are appended to iox_bench.lp

```

**Review by pauldix — APPROVED** [review]
Looks great, just one question. What timestamps are given to the data? It looks like they aren't put in so you're expecting the server to set them so they'll just be whenever the data was imported?

Can we pull the timestamps from the criterion files and use those? I think that would make it for when the actual benches were run?

**Comment by alamb** [comment]
@pauldix  -- the timestamps in the line protocol file are the timestamps of the *git commit* being tested (not, for example, the time the actual benchmarks were run). The git-fu I found on the internets for this was:

```
    TIMESTAMP=$((`git show -s --format="%ct" $GITSHA` * 1000000000)) # ct means seconds since epoch
```

I was thinking that then we could more easily correlate performance of IOx over time rather than the times that the benchmarks themselves were run

**Comment by alamb** [comment]
I am going to merge this one in -- @pauldix  if you want different timestamps other than the git commit, please just let me know.

**Comment by pauldix** [comment]
@alamb ah yes, that's good 👍

**Pull request merged**